Add FXIOS-8544  Telemetry Probes - Address Autofill Phase 1

razvanlitianu opened the pull request:
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-8544)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/18980)

## :bulb: Description
<!--- Describe your changes so the reviewer can understand the context and decisions made for your work -->

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

nbhasin2 commented on firefox-ios/Client.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:
```diff
@@ -1,14 +1,5 @@
 {
   "pins" : [
-    {
```
@razvanlitianu do we need this? seems like this got pushed accidentally 🤔

razvanlitianu replied:
Correct, SwiftPM is acting up lately. Will revert

nbhasin2 commented on firefox-ios/Client/Frontend/TabContentsScripts/FormAutofillHelper/FormAutofillHelper.swift:
```diff
@@ -18,6 +18,17 @@ struct AutofillFieldValuePayload {
 }
 
+extension WKScriptMessage {
+    func decodeBody<T: Decodable>(as type: T.Type) -> T? {
```
nit: Is this is a one off instance?

If its one time usage only in this class then we can move it to this class 
```
    // MARK: Script message decoder
    func decodeBody<T: Decodable>(as type: T.Type, body: Any) -> T? {
        guard
            let dict = (body as? [String: Any]),
            let data = try? JSONSerialization.data(withJSONObject: dict, options: []) else {
            return nil
        }
        return try? JSONDecoder().decode(type, from: data)
    }
 ```
 


 Otherwise if you plan to use it multiple times then best to move it in one of these areas
 
 <img width="248" alt="Screenshot 2024-03-18 at 2 37 11 PM" src="https://github.com/mozilla-mobile/firefox-ios/assets/8919439/88b73c74-808a-4834-b36b-e765e5b9fb5f">
<img width="272" alt="Screenshot 2024-03-18 at 2 37 38 PM" src="https://github.com/mozilla-mobile/firefox-ios/assets/8919439/cc31e335-e3d5-4faf-9cdd-ff3e8b416f35">

razvanlitianu replied:
I can move it, it's also used for cc

nbhasin2 commented on firefox-ios/Client/Frontend/TabContentsScripts/FormAutofillHelper/FillAddressAutofillForm.swift:
```diff
@@ -5,11 +5,38 @@
-    let type: String
+    let payload: Payload
+    let type: FormAutofillPayloadType
 
-    enum CodingKeys: String, CodingKey, CaseIterable {
-        case addressPayload = "payload"
-        case type = "type"
+    struct Payload: Codable {
+        var addressLevel1: String
+        var addressLevel2: String
```
Would be good to add a small test here for this codable as to make sure the payload we are expecting from formautofill is valid.

Why we need to test?
The payload can accidentally change when we import scripts or if someone accidentally pushes unwarranted changes.

razvanlitianu replied:
I agree. I think a better option would be to make them all optional since they're too impredictible. Also while we are here I noticed we have a discrepancy of naming the variables. In the telemetry callback we have ` "address_level1"` with underscore and here with hiphen  `"address-level1"`.
cc: @issammani

issammani replied:
Yes. That's one of the many annoying things I don't like about our data format. Once we move to glean we can fix this. Is it too much of a hassle to handle the different cases now in swift ? If that's the case, let's drop the extras until we move to glean on Desktop.

razvanlitianu replied:
Not at all, we just have to duplicate the decoding keys, that's all.

issammani replied:
Nice. Thanks. I will open a ticket from desktop side to match the names in the future.

adudenamedruby commented on firefox-ios/Client/metrics.yaml:
```diff
@@ -4880,7 +4880,96 @@ places_history_migration:
+  saved_all:
+    type: counter
+    description: |
+      A counter of the number of all addresses that are currently saved by the user.
```
This is cumulative, though, and we're expecting it to be cumulative, right?

razvanlitianu replied:
I think so. Up until now I noticed counters were used to represent taps or how many times you opened something. This time my understanding is to report how many addresses a user has.

adudenamedruby replied:
But, if we're adding to a counter, and the counter is at, say, 15, and you add, say, count = 3, the new counter is 18, no? And if the user removes two, so count = 1, the new counter will be 19, which is not total they have now? @travis79 am I mistaken here?

razvanlitianu replied:
Thanks for the sugestions! It seems quantity metric is the way to go, being implemented also for credit cards and logins.

adudenamedruby commented on BrowserKit/Sources/Common/Extensions/WKScriptMessageExtension.swift:
```diff
@@ -0,0 +1,16 @@
+
+public extension WKScriptMessage {
+    func decodeBody<T: Decodable>(as type: T.Type) -> T? {
+        guard
+            let dict = (body as? [String: Any]),
+            let data = try? JSONSerialization.data(withJSONObject: dict, options: []) else {
+            return nil
+        }
```
total personal nit lol
```suggestion
        guard let dict = (body as? [String: Any]),
                   let data = try? JSONSerialization.data(withJSONObject: dict, options: [])
        else { return nil }
```

nbhasin2 commented:
@razvanlitianu overall looks good with minor improvements and we are good to merge 🤞🏼 
- small test 
- metrics update (make sure to update the test if it is required)
- review from Roux when you are ready

razvanlitianu commented:
# Request for data collection review form

1) What questions will you answer with this data?
- How often are address forms detected, indicating the frequency at which users are presented with opportunities to utilize autofill functionalities?
- What is the engagement level of users with autofill prompts, including their actions to show, expand, or dismiss these prompts?

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? Some example responses:
- By examining events such as form detections, prompt actions, autofill success rates, modifications post-autofill, and user management of saved addresses, we can gain insights into how users interact with and perceive the autofill feature.

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?
- No alternatives

4) Can current instrumentation answer these questions?
- No

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   

**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**

| Event Name                                  | Description                                                                                      | Tracking Bug                                                                |
|---------------------------------------------|--------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------|
| addresses.form_detected                     | Recorded when a form is recognized as an address form.                                           | [Firefox iOS Issue #18980](https://github.com/mozilla-mobile/firefox-ios/issues/18980) |
| addresses.autofill_prompt_shown             | Recorded when autofill popup is shown, indicating which field triggered this event.              | [Firefox iOS Issue #18980](https://github.com/mozilla-mobile/firefox-ios/issues/18980) |
| addresses.autofill_prompt_expanded          | Recorded when autofill prompt was expanded.                                                      | [Firefox iOS Issue #18980](https://github.com/mozilla-mobile/firefox-ios/issues/18980) |
| addresses.autofill_prompt_dismissed         | Recorded when autofill prompt was dismissed.                                                     | [Firefox iOS Issue #18980](https://github.com/mozilla-mobile/firefox-ios/issues/18980) |
| addresses.autofilled                        | Recorded when a form is autofilled.                                                              | [Firefox iOS Issue #18980](https://github.com/mozilla-mobile/firefox-ios/issues/18980) |
| addresses.modified                          | Recorded when a field is autofilled and then modified by the user.                               | [Firefox iOS Issue #18980](https://github.com/mozilla-mobile/firefox-ios/issues/18980) |
| settings_autofill                           | Recorded when user has tapped Autofill in settings menu.                                         | [Firefox iOS Issue #18980](https://github.com/mozilla-mobile/firefox-ios/issues/18980) |
| address_form_ext_detected                   | Recorded when a form is recognized as an address form.                                           | [Firefox iOS Issue #18980](https://github.com/mozilla-mobile/firefox-ios/issues/18980) |
| address_form_ext_filled                     | Recorded when a form is autofilled. Possible values are filled, not_filled, user_filled or unavailable | [Firefox iOS Issue #18980](https://github.com/mozilla-mobile/firefox-ios/issues/18980) |
| formautofill.addresses.detected_sections_count | Count at detection time number of address form sections. A single form can contain more than one address form section. | [Firefox iOS Issue #18980](https://github.com/mozilla-mobile/firefox-ios/issues/18980) |
| addresses.deleted                           | A counter of the number of addresses that have been deleted by the user.                         | [Firefox iOS Issue #18980](https://github.com/mozilla-mobile/firefox-ios/issues/18980) |
| addresses.saved_all                         | A counter of the number of all addresses that are currently saved by user.                       | [Firefox iOS Issue #18980](https://github.com/mozilla-mobile/firefox-ios/issues/18980) |


6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
https://dictionary.telemetry.mozilla.org/apps/firefox_ios

7) How long will this data be collected?  Choose one of the following:
- Currently set to expire 2024-10-01

8) What populations will you measure?
- All channels, all locales, all countries

9) If this data collection is default on, what is the opt-out mechanism for users?
- The user can opt out

10) Please provide a general description of how you will analyze this data.
- Glean

11) Where do you intend to share the results of your analysis?
- Glean and ad partners

12) Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection? If so:
- No third party tools

adudenamedruby commented:
r+
# Data Review Form

_1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?_ 

Yes, telemetry is documented at https://dictionary.telemetry.mozilla.org/apps/firefox_ios

_2) Is there a control mechanism that allows the user to turn the data collection on and off?_ 

Yes, Firefox iOS provides data controls in the app settings

_3) If the request is for permanent data collection, is there someone who will monitor the data over time?_

This will be reviewed by a data-steward on the Firefox iOS team (@adudenamedruby) every 6 months

_4) Using the **[category system of data types](https://wiki.mozilla.org/Firefox/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?_

Type 2

_5) Is the data collection request for default-on or default-off?_

Default on

_6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?_

No

_7) Is the data collection covered by the existing Firefox privacy notice?_ 

Yes

_8) Does the data collection use a third-party collection tool?_ 

No

razvanlitianu commented:
green on buitrise build.

razvanlitianu commented:
@mergify backport release/v124

razvanlitianu commented:
@mergify backport release/v125
